src/geoip_cache.rs

```diff
@@ -0,0 +1,105 @@
+use serde::{Deserialize, Serialize};
+use std::net::IpAddr;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+use std::time::SystemTime;
+use std::{collections::HashMap, fs, io, time::Duration};
+use tokio::sync::RwLock;
+
+use ziggurat_core_geoip::geoip::{GeoIPInfo, GeoIPService};
+
+#[derive(Clone, Serialize, Deserialize)]
+struct CacheEntry {
+    pub last_updated: SystemTime,
+    pub entry: GeoIPInfo,
+}
+
+#[derive(Default, Clone, Serialize, Deserialize)]
+struct GeoCache {
+    pub entries: HashMap<IpAddr, CacheEntry>,
+}
+
+/// GeoIP cache responsible for getting and caching results.
+pub struct GeoIPCache {
+    /// Available providers and their configuration.
+    providers: Vec<Box<dyn GeoIPService>>,
+    /// Path to the cache file.
+    cache_file: PathBuf,
+    /// Cache entries.
+    cache: Arc<RwLock<GeoCache>>,
+}
+
+impl GeoIPCache {
+    /// Create a new GeoIP cache.
+    pub fn new(cache_file: &Path) -> Self {
+        Self {
+            providers: Vec::new(),
+            cache_file: cache_file.to_owned(),
+            cache: Arc::new(RwLock::new(GeoCache::default())),
+        }
+    }
+
+    /// Add a new provider to the list of providers. The providers will be called in the order they
+    /// are added.
+    pub fn add_provider(&mut self, provider: Box<dyn GeoIPService>) {
+        self.providers.push(provider);
+    }
+
+    /// Load the cache from the file.
+    pub async fn load(&self) -> Result<(), io::Error> {
+        let cache_string = fs::read_to_string(&self.cache_file)?;
```

Comment:
hmhm, we can have a check here to check the cache_file stats to see if the file is older than two weeks - if yes, we can discard it (and delete it) before we even read it here since all IPs will be too old

Reply:
I think that removing entries is sufficient for now, in case if we'd like to stick to some old records due to geo limits.